lib/internal/modules/esm/loader.js

```diff
@@ -76,12 +84,21 @@ class Loader {
         'object', 'loader resolve', resolved
       );
 
-    const { url, format } = resolved;
-
+    const { url } = resolved;
+    let { format } = resolved;
     if (typeof url !== 'string')
       throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
         'string', 'loader resolve', 'url', url
       );
+    if (this._getFormat !== undefined) {
+      ({ format } = await this._getFormat({ url, isMain, defaultGetFormat }));
+    }
+    if (!format && new URL(url).protocol === 'file:') {
+      throw new ERR_UNKNOWN_FILE_EXTENSION(
+        extname(new URL(url).pathname),
+        fileURLToPath(url),
+        fileURLToPath(parentURL));
```

Comment:
Is there a reason this can't be thrown in `getFormat` itself?

Ideally we could have an INVALID_ARG error for no `format` returned from `getFormat` (the default `getFormat` function can just be an identify function on the format input from resolve).

Reply:
Can you explain this a little more?

I wouldn’t want to require generic error handling in any user-written hook, to avoid more boilerplate that users would need to write. Or are you referring to including it in the default `getFormat`?

Reply:
It could, but the more I look at this, the more I feel that an unknown format is non-exceptional until it gets to the runtime itself.

compare:

```mjs
export getFormat(...args) {
   try {
    // data: returns null on unknown
    const format = defaultGetFormat();
    // branch for null or string
  } catch (e) {
    // branch for file extensions
    if (e.code === 'UNKNOWN_FILE_EXTENSION') { ... }
  }
}
```

```mjs
export getFormat(...args) {
  // data: returns null on unknown
  const format = defaultGetFormat(...args);
  // branch for null or string
  // branch for file extensions
}
```

I'm not sure what we gain by throwing/rejecting here really, maybe if that could be explained some on why it shouldn't be a return value of some kind (doesn't have to be `null`).

Reply:
I think the intent was to make the ESM loader limited to only known file types, so we can add more in the future and it's not like CommonJS where any unknown extension is treated as JavaScript (and therefore supporting new extensions for non JavaScript like wasm is a breaking change).

Reply:
@GeoffreyBooth for eventual evaluation, yes. We are not changing throwing on finding an unknown type, just for the hooks the delegation wouldn't throw and instead return `null`.

Reply:
* they can already return `null` to be clear

Reply:
The issue here is that this error message mentions the file extension as the issue, when the extension lookup was done as a specific implementation of the `getFormat` hook.

If we want this to be a generic error outside of the specific default `getFormat` hook, then it should be a generic error message along the lines of `"No format found for module [id]"`, as opposed to a specific one about what information (the file extension) led to the format not being detected.

Reply:
@guybedford would changing the error message be sufficient?